### DISCUSSION TITLE
The "null-forgiving" ("!") operator's naming is confusing

### DISCUSSION THREAD

**Discussion by marwalsch** [dotnet/csharplang#5161] [General]
The is might be treated as a proposal, otherwise any clarification on the intention of this particular name is welcome. Note that this might partially subjective.
When speaking of "null-forgiving", it implies something similar to `null` is allowed where it would not be otherwise". This is essentially what `Nullable<T>` and its shorthand equivalent of an appended `?` do, fulfilling the exact opposite purpose of `!`.

Therefore, i think it would be more intuitive to treat `!` as a "[non-null-assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator)"-operator
This would also avoid confusion with an appended `?` which considering its purpose has more reason of being named "null-forgiving".

**Comment by jnm2** [comment]
'Assertion' makes it sound like it is itself a runtime check, which would be misleading since it is not.

I'd rather it had been called a null-oblivious or null-blind operator. The most accurate name for it would be "nullability warning suppression operator."

**Reply by marwalsch** [reply]
So is telling the compiler to forgive null values through an operator.
The new naming proposal here might not be optimal, but the current one is even more irritating since an operation is generally something occurring at runtime. `!` serves more of a compiler directive though.
Something like "non-null-conjecture" might be more suitable.

**Reply by jnm2** [reply]
'Operator' is the only way to describe what it is syntactically.

**Reply by marwalsch** [reply]
Again, I am not trying to change it's nature or implementation, I am saying the fact that an operator used for such purpose should not add additional confusion through ambiguous naming; especially since the `Nullable<T>` shorthand is even more suitable to be labelled "null-forgiving".

**Reply by marwalsch** [reply]
"null-forgiving" labels the operator as such which forgives null values, but it is actually a directive to the compiler to forgive null values.

**Reply by jnm2** [reply]
I guess I'll start calling it the "null-blindness operator" and hope it catches on. :)

**Reply by marwalsch** [reply]
I deem "not-null-conjecture" or "null-indifference" genuinely good alternatives.

**Reply by HaloFour** [reply]
Therein lies the problem, there are probably a dozen names you could throw at them, each being slightly incorrect in one nuanced way or another.  Either way it's an operator that does one specific thing that a developer will have to learn how to use once, and as long as it's name isn't horribly wrong I don't think it's terribly necessary to bikeshed over replacements that would only create more confusion.

**Reply by marwalsch** [reply]
I don't see how any of these would create more confusion considering the current ambiguity of the name.

**Reply by yaakov-h** [reply]
I liked “the dammit operator” before it got its current name…

**Reply by marwalsch** [reply]
@yaakov-h As in "damn those compiler warnings"?

**Reply by HaloFour** [reply]
> I don't see how any of these would create more confusion considering the current ambiguity of the name.

You learn the term once and you move on.  I don't find it any more confusing than any of many other names assigned to different programming concepts that would be difficult to pin to a single name.  This feature shipped two years ago, and the operator has been discussed on these forums for some time before that, and this is the first protest regarding the name that I've seen so I would suggest that the idea that the name is confusing is certainly not universal, nor is it worth revisiting at this point.

**Reply by TonyValenti** [reply]
How about "Null Clarifying" operator?
